### PR TITLE
Enable Mixtral8x7B

### PR DESCRIPTION
**Pull request by intellinjun**
## Type of Change

feature or bug fix or documentation or others
Model enabling

## Description
Enable [Mixtral-8x7B](https://huggingface.co/mistralai/Mixtral-8x7B-Instruct-v0.1)
- Three new OP,mul_mat_id,argsort,top_k
- Two hparameter n_experts and n_experts_used
- Use GGUF format model
- Support F32 F16 Jblas

image

### CODE REVIEW

File: neural_speed/models/llama/llama_utils.cpp [hunk 1]
```diff
@@ -176,10 +192,24 @@ void Llama::load(model_context* ctx, model_progress_callback progress_callback,
       layer.norm[1] = ml->get_tensor(layers_i + ".ffn_norm.weight", {n_embd}, backend);
 
       // ffn GEMM
-      layer.ffn[0] = ml->get_tensor(layers_i + ".feed_forward.w1.weight", {n_embd, n_ff}, backend);
-      layer.ffn[1] = ml->get_tensor(layers_i + ".feed_forward.w2.weight", {n_ff, n_embd}, backend);
-      layer.ffn[2] = ml->get_tensor(layers_i + ".feed_forward.w3.weight", {n_embd, n_ff}, backend);
 
+      if (ml->verify_tensor(layers_i + ".feed_forward.w1.weight")) {
+        layer.ffn[0] = ml->get_tensor(layers_i + ".feed_forward.w1.weight", {n_embd, n_ff}, backend);
```
Comment (DDEle) [inline neural_speed/models/llama/llama_utils.cpp:197]: Assert n_expert  and n_expert_used  to be zero here?

File: neural_speed/models/llama/llama.cpp
Comment (DDEle) [inline neural_speed/models/llama/llama.cpp:1]: Some temporary debug code left.

File: neural_speed/models/llama/llama.cpp [hunk 2]
```diff
@@ -347,17 +356,64 @@ static bool llama_model_eval_internal(model_context* ctx, const model_input* inp
...
       } else {
-        struct ne_tensor* tmp = ne_mul_mat(ctx0, model.layers[il].ffn[2], cur);
-        cur = ne_mul_mat(ctx0, model.layers[il].ffn[0], cur);
-        cur = ne_silu(ctx0, cur);
-        cur = ne_mul(ctx0, cur, tmp);
-        cur = ne_mul_mat(ctx0, model.layers[il].ffn[1], cur);
+        ne_tensor* logits = ne_mul_mat(ctx0, model.layers[il].ffn_gate_inp, cur);  // [n_tokens, num_experts]
+        ne_tensor* probs = ne_soft_max_inplace(ctx0, logits);
+        ne_tensor* selected_experts = ne_top_k(ctx0, probs, n_expert_used);
+        ne_build_forward_expand(&gf, selected_experts);
```
Comment (DDEle) [inline neural_speed/models/llama/llama.cpp:375]: Unnecessary to explicitly build graph for `selected_experts` as it is an operand of `ne_get_rows` ?

File: neural_speed/convert/convert_bloom.py [hunk 3]
```diff
@@ -96,6 +96,8 @@ def main(args_in: Optional[List[str]] = None) -> None:
     fout.write(struct.pack("i", 0))  # word_embed_proj_dim (for opt)
     fout.write(struct.pack("i", 0))  # do_layer_norm_before (for opt)
 
+    fout.write(struct.pack("i", 0))
+    fout.write(struct.pack("i", 0))
```
Comment (airMeng) [inline neural_speed/convert/convert_bloom.py:100]: a lot zeros, I know they are placeholders, can you give some comments that what are they?

File: neural_speed/core/ne_layers.c [hunk 4]
```diff
@@ -6889,157 +6959,809 @@ static void ne_compute_forward_mul_mat_bias_q_f32_bestla(const struct ne_compute
...
+      for (int64_t k = j + 1; k < ne0; k++) {
+        if (src_data[dst_data[j]] < src_data[dst_data[k]]) {
+          int32_t tmp = dst_data[j];
+          dst_data[j] = dst_data[k];
+          dst_data[k] = tmp;
+        }
+      }
+    }
+  }
+}
```
Comment (airMeng) [inline neural_speed/core/ne_layers.c:7544]: you can separate this function into another file like GEMM, ```layers/reduce/argsort.cpp``` and call std::sort instead

**Review by airMeng — COMMENTED** [review]
MOE only applies to llama now?

File: neural_speed/models/llama/llama.cpp [hunk 5]
```diff
@@ -347,17 +356,64 @@ static bool llama_model_eval_internal(model_context* ctx, const model_input* inp
...
+          ne_set_name(cur_expert, "ffn_moe_down");
+
+          cur_expert =
+              ne_mul(ctx0, cur_expert,
+                     ne_repeat(ctx0, ne_view_2d(ctx0, weights, 1, N, weights->nb[1], i * weights->nb[0]), cur_expert));
+          ne_set_name(cur_expert, "ffn_moe_weighted");
+
+          if (i == 0) {
+            moe_out = cur_expert;
+            ne_build_forward_expand(&gf, moe_out);
```
Comment (DDEle) [inline neural_speed/models/llama/llama.cpp:409]: Probably it is unnecessary to explicitly build graph for `moe_out` as it is an operand of `ne_add` / used as cur later?

**Review by a32543254 — APPROVED** [review]
LGTM

File: neural_speed/core/layers/Ops.h [hunk 6]
```diff
@@ -48,6 +48,7 @@ enum ne_op {
 
   NE_OP_MUL_MAT,
   NE_OP_MUL_MAT_BIAS,
+  NE_OP_MUL_MAT_ID,
```
Comment (a32543254) [inline neural_speed/core/layers/Ops.h:51]: what the difference between NE_OP_MUL_MAT_ID and NE_OP_MUL_MAT

File: neural_speed/models/llama/llama.cpp [hunk 7]
```diff
@@ -144,8 +147,14 @@ static bool llama_model_eval_internal(model_context* ctx, const model_input* inp
 
   struct ne_tensor* embd = ne_new_tensor_1d(ctx0, NE_TYPE_I32, N, NE_SIZE_CALC);
   ne_set_name(embd, "embd");
+  // uint32_t input_tokens[5] = {1, 5713, 3714,  264,  727};
+  // for (int i = 0; i < N; ++i) {
```
Comment (a32543254) [inline neural_speed/models/llama/llama.cpp:151]: remove unuse code

**Review by DDEle — APPROVED** [review]

**Comment by intellinjun** [comment]
> MOE only applies to llama now?

MOE only applies to mixtral, mixtral is only different from  ffn compared to llama

File: neural_speed/convert/convert_bloom.py (hunk 3 above)
Comment (intellinjun) [reply neural_speed/convert/convert_bloom.py:100]: No problem

**Comment by a32543254** [comment]
Could you post some performance data for this new model here?

File: neural_speed/core/layers/Ops.h (hunk 6 above)
Comment (intellinjun) [reply neural_speed/core/layers/Ops.h:51]: NE_OP_MUL_MAT_ID has more input tensor, selected expert id and 8 expert weights, different tokens use the expert id to select different expert weights for calculation.

**Comment by intellinjun** [comment]
> Could you post some performance data for this new model here?

I will add it to ci and use ci to test performance

File: neural_speed/models/llama/llama_utils.cpp (hunk 1 above)
Comment (intellinjun) [reply neural_speed/models/llama/llama_utils.cpp:197]: OK

File: neural_speed/convert/convert_llama.py [hunk 8]
```diff
@@ -1089,6 +1089,8 @@ def write_file_header(self, params: Params, file_type: NEFileType) -> None:
         self.fout.write(struct.pack("i", params.ffn_hidden_size))
         self.fout.write(struct.pack("i", 0))
 
+        self.fout.write(struct.pack("i", 0))
+        self.fout.write(struct.pack("i", 0))
```
Comment (Zhenzhong1) [inline neural_speed/convert/convert_llama.py:1093]: please also add these parameters in  convert_quantized_llama.py and other convert_quantized_xxxx.py files.

**Comment by Zhenzhong1** [comment]
If the GGUF version is updated in the future, please add the GGUF format for the mixtral.

File: neural_speed/models/llama/llama_utils.cpp (hunk 1 above)
Comment (intellinjun) [reply neural_speed/models/llama/llama_utils.cpp:201]: done

File: neural_speed/models/llama/llama.cpp
Comment (intellinjun) [reply neural_speed/models/llama/llama.cpp:1]: done

File: neural_speed/models/llama/llama.cpp (hunk 2 above)
Comment (intellinjun) [reply neural_speed/models/llama/llama.cpp]: done

File: neural_speed/models/llama/llama.cpp (hunk 5 above)
Comment (intellinjun) [reply neural_speed/models/llama/llama.cpp]: done

File: neural_speed/models/llama/llama.cpp (hunk 7 above)
Comment (intellinjun) [reply neural_speed/models/llama/llama.cpp]: done

File: neural_speed/convert/convert_llama.py (hunk 8 above)
Comment (intellinjun) [reply neural_speed/convert/convert_llama.py]: done

**Comment by intellinjun** [comment]
https://inteltf-jenk.sh.intel.com/job/neural_speed_extension/59/

File: neural_speed/core/ne_layers.c (hunk 4 above)
Comment (intellinjun) [reply neural_speed/core/ne_layers.c:7544]: I didn't find gemm, are you talking about conv.h and conv.cpp?

Comment (airMeng) [reply neural_speed/core/ne_layers.c:7544]: yes

Comment (intellinjun) [reply neural_speed/core/ne_layers.c]: down

**Comment by intellinjun** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>


model | quant_type | input_token | output_token | cores/instance | first token | next token | total latency
-- | -- | -- | -- | -- | -- | -- | --
mixtral-8x7b | q4_j_b128 | 32 | 32 | 32 | 803.09 | 184.41 | 6519.8
mixtral-8x7b | q4_j_b128 | 32 | 32 | 48 | 836.7 | 208.27 | 7293.07
mixtral-8x7b | q4_j_b128 | 32 | 32 | 56 | 741.93 | 210.23 | 7259.06
mixtral-8x7b | q4_j_b128 | 1024 | 32 | 32 | 20802.97 | 189.49 | 26677.16
mixtral-8x7b | q4_j_b128 | 1024 | 32 | 48 | 18736 | 211.79 | 25301.49
mixtral-8x7b | q4_j_b128 | 1024 | 32 | 56 | 17763.37 | 217.67 | 24511.14
mixtral-8x7b | q4_j_b32 | 32 | 32 | 32 | 915.76 | 201.5 | 7162.26
mixtral-8x7b | q4_j_b32 | 32 | 32 | 48 | 893.81 | 215.74 | 7581.75
mixtral-8x7b | q4_j_b32 | 32 | 32 | 56 | 867.74 | 221.13 | 7722.77
mixtral-8x7b | q4_j_b32 | 1024 | 32 | 32 | 22315 | 200.47 | 28529.57
mixtral-8x7b | q4_j_b32 | 1024 | 32 | 48 | 20216.37 | 222.51 | 27114.18
mixtral-8x7b | q4_j_b32 | 1024 | 32 | 56 | 20029.26 | 230.25 | 27167.01



</body>

</html>

**Review by luoyu-intel — APPROVED** [review]

**Comment by intellinjun** [comment]
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/linjun1/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link=blue vlink=purple>


model | quant_type | input_token | output_token | cores/instance | first token | next token | total latency
-- | -- | -- | -- | -- | -- | -- | --
mixtral-8x7b | q4_j_b128 | 32 | 32 | 32 | 120.17 | 41.14 | 1395.51
mixtral-8x7b | q4_j_b128 | 32 | 32 | 48 | 139.71 | 37.54 | 1303.45
mixtral-8x7b | q4_j_b128 | 32 | 32 | 56 | 144.02 | 37.73 | 1313.65
mixtral-8x7b | q4_j_b128 | 1024 | 32 | 32 | 1408.29 | 44.63 | 2791.82
mixtral-8x7b | q4_j_b128 | 1024 | 32 | 48 | 1418.5 | 40 | 2658.5
mixtral-8x7b | q4_j_b128 | 1024 | 32 | 56 | 1301.49 | 42.21 | 2610
mixtral-8x7b | q4_j_b32 | 32 | 32 | 32 | 250.65 | 52.39 | 1874.74
mixtral-8x7b | q4_j_b32 | 32 | 32 | 48 | 391.57 | 52.54 | 2020.31
mixtral-8x7b | q4_j_b32 | 32 | 32 | 56 | 280.68 | 49.85 | 1826.03
mixtral-8x7b | q4_j_b32 | 1024 | 32 | 32 | 3414.51 | 59.43 | 5256.84
mixtral-8x7b | q4_j_b32 | 1024 | 32 | 48 | 2960.75 | 52.52 | 4588.87
mixtral-8x7b | q4_j_b32 | 1024 | 32 | 56 | 2841.19 | 51.34 | 4432.73



</body>

</html>

**Comment by intellinjun** [comment]
> <style> </style>
> model	quant_type	input_token	output_token	cores/instance	first token	next token	total latency
> mixtral-8x7b	q4_j_b128	32	32	32	120.17	41.14	1395.51
> mixtral-8x7b	q4_j_b128	32	32	48	139.71	37.54	1303.45
> mixtral-8x7b	q4_j_b128	32	32	56	144.02	37.73	1313.65
> mixtral-8x7b	q4_j_b128	1024	32	32	1408.29	44.63	2791.82
> mixtral-8x7b	q4_j_b128	1024	32	48	1418.5	40	2658.5
> mixtral-8x7b	q4_j_b128	1024	32	56	1301.49	42.21	2610
> mixtral-8x7b	q4_j_b32	32	32	32	250.65	52.39	1874.74
> mixtral-8x7b	q4_j_b32	32	32	48	391.57	52.54	2020.31
> mixtral-8x7b	q4_j_b32	32	32	56	280.68	49.85	1826.03
> mixtral-8x7b	q4_j_b32	1024	32	32	3414.51	59.43	5256.84
> mixtral-8x7b	q4_j_b32	1024	32	48	2960.75	52.52	4588.87
> mixtral-8x7b	q4_j_b32	1024	32	56	2841.19	51.34	4432.73

use mul_mat_id_silu fusion

**Comment by luoyu-intel** [comment]
> > <style> </style>
> > model	quant_type	input_token	output_token	cores/instance	first token	next token	total latency
> > mixtral-8x7b	q4_j_b128	32	32	32	120.17	41.14	1395.51
> > mixtral-8x7b	q4_j_b128	32	32	48	139.71	37.54	1303.45
> > mixtral-8x7b	q4_j_b128	32	32	56	144.02	37.73	1313.65
> > mixtral-8x7b	q4_j_b128	1024	32	32	1408.29	44.63	2791.82
> > mixtral-8x7b	q4_j_b128	1024	32	48	1418.5	40	2658.5
> > mixtral-8x7b	q4_j_b128	1024	32	56	1301.49	42.21	2610
> > mixtral-8x7b	q4_j_b32	32	32	32	250.65	52.39	1874.74
> > mixtral-8x7b	q4_j_b32	32	32	48	391.57	52.54	2020.31
> > mixtral-8x7b	q4_j_b32	32	32	56	280.68	49.85	1826.03
> > mixtral-8x7b	q4_j_b32	1024	32	32	3414.51	59.43	5256.84
> > mixtral-8x7b	q4_j_b32	1024	32	48	2960.75	52.52	4588.87
> > mixtral-8x7b	q4_j_b32	1024	32	56	2841.19	51.34	4432.73
> 
> use mul_mat_id_silu fusion

It's SiLu that slows down the first-token inference, right?

**Review by airMeng — APPROVED** [review]

**Comment by intellinjun** [comment]
> > > <style> </style>
> > > model	quant_type	input_token	output_token	cores/instance	first token	next token	total latency
> > > mixtral-8x7b	q4_j_b128	32	32	32	120.17	41.14	1395.51
> > > mixtral-8x7b	q4_j_b128	32	32	48	139.71	37.54	1303.45
> > > mixtral-8x7b	q4_j_b128	32	32	56	144.02	37.73	1313.65
> > > mixtral-8x7b	q4_j_b128	1024	32	32	1408.29	44.63	2791.82
> > > mixtral-8x7b	q4_j_b128	1024	32	48	1418.5	40	2658.5
> > > mixtral-8x7b	q4_j_b128	1024	32	56	1301.49	42.21	2610
> > > mixtral-8x7b	q4_j_b32	32	32	32	250.65	52.39	1874.74
> > > mixtral-8x7b	q4_j_b32	32	32	48	391.57	52.54	2020.31
> > > mixtral-8x7b	q4_j_b32	32	32	56	280.68	49.85	1826.03
> > > mixtral-8x7b	q4_j_b32	1024	32	32	3414.51	59.43	5256.84
> > > mixtral-8x7b	q4_j_b32	1024	32	48	2960.75	52.52	4588.87
> > > mixtral-8x7b	q4_j_b32	1024	32	56	2841.19	51.34	4432.73
> > 
> > 
> > use mul_mat_id_silu fusion
> 
> It's SiLu that slows down the first-token inference, right?

Silu slows down the next token inference ,first token is due to a problem with the thread setup in mul_mat_id

File: neural_speed/models/llama/llama.cpp [hunk 9]
```diff
@@ -351,17 +354,67 @@ static bool llama_model_eval_internal(model_context* ctx, const model_input* inp
...
+        ne_tensor* weights = ne_get_rows(ctx0, ne_reshape_3d(ctx0, probs, 1, n_expert, N), selected_experts);
+        weights = ne_reshape_2d(ctx0, weights, n_expert_used, N);
+        ne_tensor* weights_sum = ne_sum_rows(ctx0, weights);
+        weights_sum = ne_repeat(ctx0, weights_sum, weights);
+        weights = ne_div(ctx0, weights, weights_sum);
+        ne_tensor* moe_out = nullptr;
+
+        for (int i = 0; i < n_expert_used; ++i) {
+          ne_tensor* cur_expert;
+          if (N == 1 && ctx0, model.layers[il].ffn_down_exp[0]->type == NE_TYPE_BTLA) {
```
[the remark this answers is not in the document]
Comment (intellinjun) [reply neural_speed/models/llama/llama.cpp]: done

**Pull request merged**